### ISSUE TITLE
How to specialise on JSON deserialiser

### ISSUE THREAD

**Issue by aldanor** [serde-rs/json#193]
Now that there's a non-public `DeserializerImpl` which is the one that's actually being used in `from_trait()`, it seems to have become impossible to use specialisation on JSON decoders when using the top-level `from_*()` methods.

E.g., I naively assumed something like this would work:

```rust
#![feature(specialization)]

struct Foo;

impl serde::Deserialize for Foo {
    fn deserialize<D>(de: &mut D) -> Result<Foo, D::Error>
        where D: serde::Deserializer
    {
        type Result<E> = Result<Foo, E>;

        trait Deserialize<D: serde::Deserializer> {
            fn deserialize(de: &mut D) -> Result<D::Error>;
        }

        impl<D: serde::Deserializer> Deserialize<D> for Foo {
            default fn deserialize(de: &mut D) -> Result<D::Error> {
                /* Default impl */
            }
        }

        impl<Iter> Deserialize<json::Deserializer<Iter>> for Foo
            where Iter: Iterator<Item=::std::io::Result<u8>>
        {
            fn deserialize(de: &mut json::Deserializer<Iter>) -> Result<json::Error> {
                /* JSON specialization */
            }
        }

        <Foo as Deserialize<D>>::deserialize(de)
    }
}
```

... and it kind of does, but the default impl is triggered when you use functions like `from_str()`, because `serde_json` doesn't use `serde_json::Deserializer` but instead a custom hidden deserialiser.

Would it make sense for `from_trait()` to somehow use `serde_json::Deserializer` instead? (apologies if I'm missing something obvious; maybe this problem is currently unavoidable)

**Comment by dtolnay** [comment]
You're right, this is not possible with the current serde_json API. We did some contortions when we introduced the DeserializeImpl-related optimizations (the separate specializations for &str vs &[u8] vs io::Read) to avoid breaking backward compatibility. I will try to rework some of this for 0.9.0 to make specialization possible.

**Comment by aldanor** [comment]
Hmm. Would it be possible to make `DeserializerImpl` public for the time being so there's at least some way to specialise on it, even it's to be scrapped later on?

**Comment by aldanor** [comment]
@dtolnay Ok, scratch the previous comment, I've found a pretty hilarious way of specialising which actually works... on the error type 😃 

As you said, would be sure nice to be able to do it without such hacks.

```rust
impl serde::Deserialize for Foo {
    fn deserialize<D>(de: &mut D) -> ::std::result::Result<Foo, D::Error>
        where D: serde::Deserializer
    {
        type Result<E> = ::std::result::Result<Foo, E>;

        trait Deserialize<D: serde::Deserializer, E: serde::de::Error> {
            fn deserialize(de: &mut D) -> Result<D::Error>;
        }

        impl<D: serde::Deserializer, E: serde::de::Error> Deserialize<D, E> for Foo {
            default fn deserialize(de: &mut D) -> Result<D::Error> {
                /* Default impl */
            }
        }

        impl<D: serde::de::Deserializer> Deserialize<D, json::Error> for Foo {
            fn deserialize(de: &mut D) -> Result<D::Error> {
                /* JSON specialization */
            }
        }

        <Foo as Deserialize<D, D::Error>>::deserialize(de)
    }
}
```

**Comment by dtolnay** [comment]
I have a fix in #205 that makes it so all deserialization goes through the public serde_json::Deserializer.
